### PR TITLE
21 create basic hd108 example

### PR DESCRIPTION
**Pull request by CodingInGreen**
Added HD108 driver using embedded_hal_async::spi::SpiDevice

But esp-hal wants spi implemented with esp_hal::spi::master::Spi

So I tinkered with creating a wrapper, but don't know exactly how to make that work.

### CODE REVIEW

File: firmware/f1-hardware/src/main.rs [hunk 1]
```diff
@@ -1,150 +1,67 @@
...
+    let system = SystemControl::new(peripherals.SYSTEM);
+    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
+
+    // Initialize the SPI interface
+    let io = Io::new(peripherals.GPIO, peripherals.IO_MUX);
+    
+    let sclk = io.pins.gpio6;
+    let miso = io.pins.gpio7;
+    let mosi = io.pins.gpio13;
+    let cs = io.pins.gpio10;
```
Comment (ijager) [inline firmware/f1-hardware/src/main.rs:38]: Probably need to configure pins for SPI mode / or alt mode (not sure how its called in ESP)

File: firmware/f1-hardware/src/main.rs [hunk 2]
```diff
@@ -1,150 +1,67 @@
...
+    .with_pins(Some(sclk), Some(mosi), Some(miso), Some(cs));
 
-    loop {
-        if stack.is_link_up() {
-            break;
-        }
-        Timer::after(Duration::from_millis(500)).await;
-    }
+    // Wrap the SPI interface
+    let spi_wrapper = SpiWrapper::new(spi);
```
Comment (ijager) [inline firmware/f1-hardware/src/main.rs:49]: Why? can't we pass in the original spi to the HD108 driver?

Comment (CodingInGreen) [reply firmware/f1-hardware/src/main.rs:49]: When I try to implement like in the embassy_spi.rs example:

error[E0277]: the trait bound `esp_hal::spi::master::Spi<'_, esp_hal::peripherals::SPI2, FullDuplexMode>: SpiDevice` is not satisfied
  --> src/main.rs:48:32
   |
48 |     let mut hd108 = HD108::new(spi);
   |                     ---------- ^^^ the trait `SpiDevice` is not implemented for `esp_hal::spi::master::Spi<'_, esp_hal::peripherals::SPI2, FullDuplexMode>`
   |                     |
   |                     required by a bound introduced by this call
   |
   = help: the trait `SpiDevice<Word>` is implemented for `&mut T`
note: required by a bound in `HD108::<SPI>::new`
  --> src/hd108.rs:10:10
   |
10 |     SPI: SpiDevice,
   |          ^^^^^^^^^ required by this bound in `HD108::<SPI>::new`
11 | {
12 |     pub fn new(spi: SPI) -> Self {
   |            --- required by a bound in this associated function

**Comment by CodingInGreen** [comment]
Pushed attempt at writing wrapper using:

https://docs.embassy.dev/embassy-embedded-hal/git/default/adapter/struct.BlockingAsync.html#impl-SpiBus-for-BlockingAsync%3CT%3E

**Comment by CodingInGreen** [comment]
esp-hal triaits:

esp-hal::spi::master::Spi (non-async)
esp-hal::spi::SpiMode (non-async)

embedded-hal traits that esp-hal::spi::master::Spi implements:

[sync]
DuplexMode
FullDuplexMode
HalfDuplexReadWrite
WithDmaSpi2
WithDmaSpi3
InstanceDma
Instance

[async]
SpiDma implements the SpiBus trait from embedded_hal_async.
The trait methods (read, write, transfer, transfer_in_place, and flush) are all asynchronous (async fn).

In the esp-hal library, the dma module defines traits like WithDmaSpi2 and WithDmaSpi3, which are implemented for SpiDma. This allows for SPI instances to be extended with DMA capabilities.
The InstanceDma trait is used to define methods for starting DMA transfers, configuring the DMA controller, and handling data buffers.

esp32c3 and DMA:
SPI with DMA: The ESP32-C3's SPI peripherals can leverage DMA to handle large data transfers.

embedded_hal_async::spi::SpiBus;

Wrapping embedded_hal_async::spi::SpiBus for use with esp-hal's SPI traits presents challenges, primarily due to differences in asynchronous programming models, DMA handling, and trait compatibility.

Using embassy-embedded-hal to wrap the SpiBus trait for async compatibility with esp-hal's SPI traits is a possible workaround. This integration allows you to leverage the asynchronous capabilities provided by embassy-embedded-hal while utilizing the hardware abstraction and DMA features of esp-hal.

**Comment by CodingInGreen** [comment]
I found this discussion on how async traits can't be implemented on esp-hal types, so the workaround is to create wrappers?

https://github.com/esp-rs/esp-hal/discussions/126

**Comment by CodingInGreen** [comment]
I pushed the most recent change, I'm getting linker errors that I'll look closely at tomorrow.

File: firmware/f1-hardware/src/hd108.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    pub async fn write_byte(&mut self, word: u8) -> Result<(), SPI::Error> {
+        self.spi.write(&[word]).await
+    }
+
+    pub async fn write_bytes(&mut self, words: &[u8]) -> Result<(), SPI::Error> {
+        self.spi.write(words).await
+    }
+
+    pub async fn transfer<'w>(
```
Comment (ijager) [inline firmware/f1-hardware/src/hd108.rs:64]: These three functions can be removed since they're not used

**Review by ijager — APPROVED** [review]

**Pull request merged**